### ISSUE TITLE
Strange behaviour in hop-by-hop cancel propagation when using MI t_reply, 2.3 branch

### ISSUE THREAD

**Issue by sisoftrg** [OpenSIPS/opensips#1337]
I have not very usual task: drop incoming 18x replies on INVITE and instead generate 18x replies by request from external application. I use `drop()` in `onreply_route` to stop forwarding of original 18x replies and tm MI command `t_reply` to build new replies. All works well except one situation: when after generating fake 18x reply UAC side sends CANCEL, it replies with `200 canceling` and not forwards to UAS.
Example:
```
UAC          OpenSIPS     UAS
INVITE ->
          <- 100 Trying
               INVITE ->
                        <- 100 Trying
               drop()   <- 183 Progress
180 Ringing <- t_reply
CANCEL ->
          <- 200 canceling
                                <==  here should be CANCEL to UAS, but it absent
                        <- 200 OK
```

I examine the code of tm module and in `t_reply.c:_reply_light()` found line:
```c
if ( is_invite(trans) ) which_cancel(trans, &cancel_bitmap );
```
`which_cancel()` internally uses `should_cancel_branch()` which have side effect: mark cancel buffer as busy on first usage. I think because of this when real CANCEL arrives, it not propagates - branch not marks to canceling.

Adding additional conditions solves above described problem in my case:
```c
if ( is_invite(trans) && code != 180 && code != 183) which_cancel(trans, &cancel_bitmap );
```

**Comment by bogdan-iancu** [comment]
Hi @sisoftrg , actually the issue is here -> https://github.com/OpenSIPS/opensips/blob/5df9679c09ea8a247baac9e9ffdaff8e238feb4a/modules/tm/t_cancel.c#L71 . As you do drop for the incoming 180, the last_received field probably does not get set, so the cancel will see that branch as an empty one (with no incoming reply), so with nothing to cancel.
I will check more.

**Comment by sisoftrg** [comment]
No, it looks not correct. When I only drop incoming 18x and not use t_reply then all works fine.
Problem occurs only when I send provisional 18x via t_reply before 200, not depending on drop/pass of original 18x.

**Comment by bogdan-iancu** [comment]
you say that the not the dropping 180 is the trigger (for the failing cancel), but the usage of t_reply MI command ?

**Comment by sisoftrg** [comment]
Yes, exactly.

**Comment by bogdan-iancu** [comment]
@sisoftrg , thanks for the info, I just committed a fix on master branch - could you please validate the fix?

Many thanks, Bogdan

**Comment by sisoftrg** [comment]
It looks correct, worked for 2.3 branch.
